File: app/src/main/java/com/zoneanchor/AlarmReceiver.java
package com.zoneanchor;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

public final class AlarmReceiver extends BroadcastReceiver {
    public static final String ACTION_FIRE_ALARM = "com.zoneanchor.action.FIRE_ALARM";
    public static final String EXTRA_ALARM_ID = "com.zoneanchor.extra.ALARM_ID";

    @Override
    public void onReceive(Context context, Intent intent) {
        int alarmId = intent.getIntExtra(EXTRA_ALARM_ID, ZonedAlarm.DEFAULT_ID);
        ZonedAlarm alarm = AlarmStore.load(context, alarmId);
        if (alarm == null || !alarm.enabled) {
            return;
        }

        NotificationHelper.showAlarm(context, alarm);
        AlarmScheduler.scheduleNext(context, alarm);
    }
}

File: app/src/main/java/com/zoneanchor/alarm/AlarmReceiver.kt
package com.zoneanchor.alarm

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.zoneanchor.ZoneAnchorApp
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class AlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val pending = goAsync()
        scope.launch {
            try {
                val app = context.applicationContext as ZoneAnchorApp
                val alarmId = intent.getIntExtra(EXTRA_ALARM_ID, com.zoneanchor.model.ZonedAlarm.DEFAULT_ID)
                val alarm = app.alarmRepository.find(alarmId)
                if (alarm != null && alarm.enabled) {
                    NotificationHelper.showAlarm(context, alarm)
                    app.alarmScheduler.scheduleNext(alarm)
                }
            } finally {
                pending.finish()
            }
        }
    }

    companion object {
        const val ACTION_FIRE_ALARM = "com.zoneanchor.action.FIRE_ALARM"
        const val EXTRA_ALARM_ID = "com.zoneanchor.extra.ALARM_ID"
        private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    }
}
